### ISSUE TITLE
FontForge should use UFO as saving format

### ISSUE THREAD

**Issue by Pecita** [fontforge/fontforge#3219]
FontForge should use UFO as saving format instead of SFD.
Many script functions would then become obsolete because it would be easier to modify the UFO source. It would also make it possible to modify the fonts with other tools and thus to open many possibilities.

**Comment by ctrlcctrlv** [comment]
UFO is already an available save format.

**Comment by frank-trampe** [comment]
Yes, @Pecita, @ctrlcctrlv is correct. We added this functionality a few years ago, including accompanying feature files. It is available from the Generate menu.

**Comment by Pecita** [comment]
This is not correct because this feature has many, many bugs and is not really usable. In addition it would take UFO 3.
I'm talking about replacing the sfd format: we are far from it!

**Comment by frank-trampe** [comment]
@Pecita, your issue doesn't really address your primary concern then. Maybe bump any UFO bugs that you've filed so that I can look at them.

We currently have UFO 3 read support. We had UFO 3 support for a while, but we disabled it because it was mangling misformed UFO 2 files, which are surprisingly common. I suppose I can add it as a separate save option.

Even if we fixed all of those, I don't see why we would replace SFD. It's a compact, flexible way of storing the entire FontForge state for a project, and a lot of our users would miss those benefits if we were to take them away.

**Comment by Pecita** [comment]
@frank-trampe, You cannot directly modify an sfd file: you have to use fontforge functions. You can edit an UFO file directly or using a personal tool or using a third-party software. A TTX format could as well be appropriate if it became popular. In fact, it's UFO or TTX.
For example, regarding my main need that is to create in batch a calt lookup, it is easy by generating an Adobe feature file while I have no way to do it by FontForge which has no function for that and whose UFO import / export is bugged . So I use makeotf with other disadvantages. I have not watched TTX yet. TTX may be an even better solution, I do not know. I believe that FontTools provide libraries that FontForge could use.
Let's be serious, in its state of complexity, without a George Williams, FontForge is sinking. Simplifying the architecture by looking for standards would keep an open font editor. UFO seems a good way.

**Comment by frank-trampe** [comment]
FontForge imports and exports feature files. I don't understand what else you need.

SFD is a serialized, standardized representation of FontForge's native data structures. UFO is a very different thing. Using it as a native format would either drop a lot of state or introduce a lot of complexity and odd edge cases (some of which we've already needed to handle implementing the present level of support).

If there are specific issues with the UFO import or export, let us know. I believe that, as currently written, even though the export functionality isn't exposed, we round-trip UFO 3. So if you change the version tag in a file before you import it, it ought to stick in the export.

**Comment by frank-trampe** [comment]
Oh. Actually, the UFO3 thing is more complicated than that. We flag version 2 but follow the example of Adobe in including layercontents.plist and any layers that either have content or were present in an imported UFO. We also export group kerning. Other tools ought to be able to read the UFO3 data even though the version is flagged as 2. We can add an option to flag the version as 3 if that would be helpful. Or, as you mention, it's fairly easy to change it by hand.

We implemented UFO 3 before anybody else supported it, and it caused a lot of compatibility problems, which is why we changed back to UFO 2 for the version flag.

**Comment by Pecita** [comment]
> FontForge imports and exports feature files. I don't understand what else you need.
For example...
1) extract the following tar.
2) open PecitaL-Regular.ufo with FF (say OK to warning!)
3) save as PecitaL-Regular1.ufo
4) close FF
5) compare the two features.fea.
Note that the first ufo file generate a valid otf font using either makeotf or FF (I worked a lot for that compatibility).
The second ufo file generate a corrupted otf file with FF.
[PecitaL-Regular.zip](https://github.com/fontforge/fontforge/files/1560011/PecitaL-Regular.zip)

**Comment by frank-trampe** [comment]
Okay. I'll take a look today or tomorrow.

**Comment by Pecita** [comment]
Another strange behavior is that when I generate a CFF font from the UFO font I lose the values:
```
<StdHW value = "60" />
<StdVW value = "60" />
<StemSnapH value = "60 65 70 150 170 220 230 240 470" />
<StemSnapV value = "60" />
```
in the CFF table.
I have to hand them over using TTX.

**Comment by frank-trampe** [comment]
@Pecita, can you send me your file?

**Comment by frank-trampe** [comment]
Or is that the same one?

**Comment by ctrlcctrlv** [comment]
>You cannot directly modify an sfd file

Yes you can. I have before. It's plaintext.

**Comment by frank-trampe** [comment]
@Pecita, it's been a busy week with Christmas parties and such, and I won't get to this until next week. It's on my list, though.
